examples_tests/41.VisibilityBuffer/main.cpp

```diff
@@ -3,815 +3,437 @@
-    assert(!meshes_bundle.isEmpty());
-    auto mesh = meshes_bundle.getContents().first[0];
+    //assert(!meshes_bundle.isEmpty());
+    auto mesh = meshes_bundle.getContents().begin()[0];
     auto mesh_raw = static_cast<asset::ICPUMesh*>(mesh.get());
 
-    core::vector<SInstanceData> instanceDataSSBOContents;
-    instanceDataSSBOContents.reserve(mesh_raw->getMeshBufferCount());
-    core::vector<uint32_t> instanceIndexUBOContents;
-    instanceIndexUBOContents.reserve(mesh_raw->getMeshBufferCount()*(16u/sizeof(uint32_t)));
+    //saving cache to file
+    qnc->saveCacheToFile<asset::EF_A2B10G10R10_SNORM_PACK32>(fs, "../../tmp/normalCache101010.sse");
 
-    //modifying push constants and default fragment shader for VT
-    for (uint32_t i = 0u; i < mesh_raw->getMeshBufferCount(); ++i)
-    {
-        SInstanceData instanceData;
-        memset(instanceData.map_data, 0xff, sizeof(instanceData.map_data));
-        instanceData.extra = 0u;
-
-        auto* mb = mesh_raw->getMeshBuffer(i);
-        auto* ds = mb->getAttachedDescriptorSet();
-        _NBL_DEBUG_BREAK_IF(!ds);
-        for (uint32_t k = 0u; k < TEX_OF_INTEREST_CNT; ++k)
-        {
-            uint32_t j = texturesOfInterest[k];
-
-            auto* view = static_cast<asset::ICPUImageView*>(ds->getDescriptors(j).begin()->desc.get());
-            auto* smplr = ds->getLayout()->getBindings().begin()[j].samplers[0].get();
-            const auto uwrap = static_cast<asset::ISampler::E_TEXTURE_CLAMP>(smplr->getParams().TextureWrapU);
-            const auto vwrap = static_cast<asset::ISampler::E_TEXTURE_CLAMP>(smplr->getParams().TextureWrapV);
-            const auto borderColor = static_cast<asset::ISampler::E_TEXTURE_BORDER_COLOR>(smplr->getParams().BorderColor);
-            auto img = view->getCreationParameters().image;
-            auto extent = img->getCreationParameters().extent;
-            if (extent.width <= 2u || extent.height <= 2u)//dummy 2x2
-                continue;
-            STextureData texData;
-            auto found = VTtexDataMap.find(img);
-            if (found != VTtexDataMap.end())
-                texData = found->second;
-            else {
-                auto imgToPack = createPoTPaddedSquareImageWithMipLevels(img.get(), uwrap, vwrap);
-                const asset::E_FORMAT fmt = imgToPack->getCreationParameters().format;
-                texData = getTextureData(imgToPack.get(), vt.get(), uwrap, vwrap, borderColor);
-                VTtexDataMap.insert({img,texData});
-            }
-
-            static_assert(sizeof(texData)==sizeof(instanceData.map_data[0]), "wrong reinterpret_cast");
-            instanceData.map_data[k] = reinterpret_cast<uint64_t*>(&texData)[0];
-        }
+    //TODO: change it to vector of smart pointers
+    core::vector<ICPUMeshBuffer*> meshBuffers;
+    for (uint32_t i = 0u; i < mesh_raw->getMeshBufferVector().size(); i++)
+        meshBuffers.push_back(mesh_raw->getMeshBufferVector()[i].get());
```

Comment:
`getMeshBufferVector` is for doing mutable things, I'd rather use `getMeshBuffers` unless you have a reason not to. Btw now you can do
```cpp
core::vector<core::smart_refctd_ptr<const ICPUMeshBuffer>> meshBuffers;
for (const auto& mb : mesh_raw->getMeshBuffers())
   meshBuffers.push_back(core::smart_refctd_ptr(mb));
```